File: bench/compare/go/graph_reach.go
package main

func graphIndex(nodes, from, to int) int {
	return from*nodes + to
}

func graphAddEdge(graph []int64, nodes, from, to int) {
	graph[graphIndex(nodes, from, to)] = 1
}

func graphHasEdge(graph []int64, nodes, from, to int) int64 {
	return graph[graphIndex(nodes, from, to)]
}

func buildGraph(nodes int) []int64 {
	graph := make([]int64, nodes*nodes)
	for i := 0; i < nodes-1; i++ {
		graphAddEdge(graph, nodes, i, i+1)
		if i+7 < nodes {
			graphAddEdge(graph, nodes, i, i+7)
		}
	}
	return graph
}

func graphReachableCount(graph []int64, nodes, start int) int64 {
	seen := make([]int64, nodes)
	queue := make([]int, 0, nodes)
	seen[start] = 1
	queue = append(queue, start)
	head := 0
	for head < len(queue) {
		node := queue[head]
		head++
		for next := 0; next < nodes; next++ {
			if graphHasEdge(graph, nodes, node, next) == 1 && seen[next] == 0 {
				seen[next] = 1
				queue = append(queue, next)
			}
		}
	}
	var total int64
	for _, mark := range seen {
		total += mark
	}
	return total
}

func main() {
	nodes := 160
	graph := buildGraph(nodes)
	var total int64
	for start := 0; start < nodes; start++ {
		total += graphReachableCount(graph, nodes, start)
	}
	if total != 12880 {
		panic(total)
	}
}

File: bench/compare/rust/graph_reach.rs
fn graph_index(nodes: usize, from: usize, to: usize) -> usize {
    from * nodes + to
}

fn graph_add_edge(graph: &mut [i64], nodes: usize, from: usize, to: usize) {
    graph[graph_index(nodes, from, to)] = 1;
}

fn graph_has_edge(graph: &[i64], nodes: usize, from: usize, to: usize) -> i64 {
    graph[graph_index(nodes, from, to)]
}

fn build_graph(nodes: usize) -> Vec<i64> {
    let mut graph = vec![0_i64; nodes * nodes];
    for i in 0..nodes - 1 {
        graph_add_edge(&mut graph, nodes, i, i + 1);
        if i + 7 < nodes {
            graph_add_edge(&mut graph, nodes, i, i + 7);
        }
    }
    graph
}

fn graph_reachable_count(graph: &[i64], nodes: usize, start: usize) -> i64 {
    let mut seen = vec![0_i64; nodes];
    let mut queue = Vec::with_capacity(nodes);
    seen[start] = 1;
    queue.push(start);
    let mut head = 0;
    while head < queue.len() {
        let node = queue[head];
        head += 1;
        for next in 0..nodes {
            if graph_has_edge(graph, nodes, node, next) == 1 && seen[next] == 0 {
                seen[next] = 1;
                queue.push(next);
            }
        }
    }
    seen.iter().copied().sum()
}

fn main() {
    let nodes = 160_usize;
    let graph = build_graph(nodes);
    let mut total = 0_i64;
    for start in 0..nodes {
        total += graph_reachable_count(&graph, nodes, start);
    }
    if total != 12880 {
        panic!("{}", total);
    }
}
